#30 samsungbackend close workaround

al-mahi opened the pull request:
driverWrapper.close() function call ends up calling external.cc files tflite::UnloadLibrary(handle) function which is essentially a dlclose() call on libsamsungbackend.so file. Moving the workaround of commenting driverWrapper.close() to deeper level is more specific to the issue. Also allocated memory dynamically for the backend object.

anhappdev commented:
@al-mahi Please make sure all tests are passed. `Format & lint / C++` is failed.

anhappdev commented on android/cpp/backends/external.h:
```diff
@@ -31,7 +31,7 @@ struct BackendFunctions {
 
   ~BackendFunctions() {
-    if (handle) {
+    if (false) {
```
This doesn't look right to me.

al-mahi replied:
@anhappdev I am basically stopping tflite::SharedLibrary::UnLoadLibrary(handle); or dlclose(handle); call here as a workaround of the app crash.

anhappdev replied:
1. Yes, `if (false)` means the code in the if block will never be called, why not just remove it?
2. Does this change have an impact on other backends? I am not familiar with the backend code yet, but the code was there for a reason. So other engineers need to review this.

al-mahi commented:
@relja128  I have updated the PR please review again.
@mohitmundhragithub Would you please let us know if you find any issue while testing with Qualcomm backend.
@freedomtan  Would you please let us know if you find any issue while testing with MTK backend.
`// handle = tflite::SharedLibrary::LoadLibrary(lib_path.c_str()); <-- master branch`
`handle = dlopen(lib_path.c_str(), RTLD_LAZY | RTLD_NODELETE | RTLD_LOCAL);<-- potential fix or workaround`

Some of the changes were still useful from the PR like dynamic memory allocation so I have not created a new PR.  @anhappdev was concerned about the changes in exeternal.h file so I reverted it back and now I added a flag while opening the backend which prevent crash for samsung backend.

d-uzlov reviewed (changes requested):
Using `RTLD_NODELETE` is [equivalent](https://linux.die.net/man/3/dlopen) to deleting `dlclose`/`UnLoadLibrary`.
I guess the desired effect here is to skip reinitialization of static variables, or something?

Removing/disabling `dlclose` does not look good in my opinion.
Code quality aside, we will keep resources from all backends in memory, because all of them will be loaded to call `mlperf_backend_matches_hardware` but never unloaded. This can affect memory usage if some backends have large static variables, or this can potentially affect other things if some backends have non-trivial static variables.

We should either change the logic to "load the matched backend once and keep the handle forever" (which can be discussed separately) or fix the issue with static variables.

If you want to use a workaround of never reloading the backend, Samsung backend should be able to just `dlopen` itself from the inside (then after app does a `dlclose` there will still be open handles, so backend will not be unloaded). Our API doesn't pass library path to backend so you can't call `dlopen` directly, but you can use [dladdr](https://man7.org/linux/man-pages/man3/dladdr.3.html) to get path.

mohitmundhragithub commented:
> @relja128 I have updated the PR please review again. @mohitmundhragithub Would you please let us know if you find any issue while testing with Qualcomm backend. @freedomtan Would you please let us know if you find any issue while testing with MTK backend. `// handle = tflite::SharedLibrary::LoadLibrary(lib_path.c_str()); <-- master branch` `handle = dlopen(lib_path.c_str(), RTLD_LAZY | RTLD_NODELETE | RTLD_LOCAL);<-- potential fix or workaround`
> 
> Some of the changes were still useful from the PR like dynamic memory allocation so I have not created a new PR. @anhappdev was concerned about the changes in exeternal.h file so I reverted it back and now I added a flag while opening the backend which prevent crash for samsung backend.

it seems to be working fine for us.

mohitmundhragithub commented on mobile_back_samsung/cpp/eden_mlperf_c_backend/jni/edenai.cc:
```diff
@@ -1050,15 +1050,17 @@ void mlperf_backend_clear(mlperf_backend_ptr_t backend_ptr) {
     const char* model_path, mlperf_backend_configuration_t* configs,
     const char* native_lib_path) {
-  Backend* samsung_backend = &ss_backend;
+  ss_backend = new Backend();
```
where do we clear this allocation?

al-mahi replied:
In line#1053 `delete samsung_backend`;

sb-huh commented on android/cpp/backends/external.cc:
```diff
@@ -79,7 +79,7 @@ BackendFunctions::BackendFunctions(const std::string& lib_path) {
     handle = tflite::SharedLibrary::LoadLibrary(wide_lib_path.c_str());
 #else
-    handle = tflite::SharedLibrary::LoadLibrary(lib_path.c_str());
+    handle = dlopen(lib_path.c_str(), RTLD_LAZY | RTLD_LOCAL | RTLD_NODELETE);
```
In my opinion, this modification looks so risky, and narrow scoped.

sb-huh commented on mobile_back_samsung/cpp/eden_mlperf_c_backend/jni/edenai.cc:
```diff
@@ -1050,15 +1050,17 @@ void mlperf_backend_clear(mlperf_backend_ptr_t backend_ptr) {
 }
 
-Backend ss_backend;
+static Backend* ss_backend;
```
I don't understand why this declaration is needed. What is the purpose of this declaration?

Do you mean that ss_backend should hold its address of Backend for its entire life cycle?

relja128 commented:
In the meeting we decided not to change unload behavior for all backends. Sticking with if Samsung is preferable to a wider change.
